product_pack/models/product_product.py

```diff
@@ -0,0 +1,117 @@
+
+        no_packs = (self | self.mapped('pack_line_ids.product_id')) - packs
+        return packs, no_packs
+
+    @api.multi
+    def price_compute(self, price_type, uom=False, currency=False,
+                      company=False):
+        packs, no_packs = self.separete_pack_products()
+        prices = super(ProductProduct, no_packs).price_compute(
+            price_type, uom, currency, company)
+        for product in packs.with_context(prefetch_fields=False):
+            pack_price = 0.0
+            for pack_line in product.pack_line_ids:
+                product_line_price = pack_line.product_id.price * (
+                    1 - (pack_line.discount or 0.0) / 100.0)
+                pack_price += (product_line_price * pack_line.quantity)
+            prices[product.id] = pack_price
+        return prices
+
+    @api.depends('list_price', 'price_extra')
+    def _compute_product_lst_price(self):
+        packs, no_packs = self.separete_pack_products()
+        super(ProductProduct, no_packs)._compute_product_lst_price()
+
+        to_uom = None
+        if 'uom' in self._context:
+            to_uom = self.env['uom.uom'].browse([self._context['uom']])
+        for product in packs:
+            list_price = product.price_compute('list_price').get(product.id)
+            if to_uom:
+                list_price = product.uom_id._compute_price(
+                    list_price, to_uom)
+            product.lst_price = list_price + product.price_extra
+
+    def _compute_quantities_dict(
+            self, lot_id, owner_id, package_id,
+            from_date=False, to_date=False):
+        res = super()._compute_quantities_dict(
+            lot_id, owner_id, package_id, from_date=from_date, to_date=to_date)
+        packs = self.filtered('pack')
+        for product in packs.with_context(prefetch_fields=False):
+            pack_qty_available = []
+            pack_virtual_available = []
+            for subproduct in product.pack_line_ids:
+                subproduct_stock = subproduct.product_id
+                sub_qty = subproduct.quantity
+                if sub_qty:
+                    pack_qty_available.append(math.floor(
+                        subproduct_stock.qty_available / sub_qty))
+                    pack_virtual_available.append(math.floor(
+                        subproduct_stock.virtual_available / sub_qty))
+            # TODO correctly calculate pack virtual available for negatives
+            res[product.id] = {
+                'qty_available': (
+                    pack_qty_available and min(pack_qty_available) or False),
+                'incoming_qty': 0,
+                'outgoing_qty': 0,
+                'virtual_available': (
+                    pack_virtual_available and
+                    max(min(pack_virtual_available), 0) or False),
+            }
+        return res
+
+    @api.constrains('pack_line_ids')
+    def check_recursion(self):
```

Comment:
Move this check recursion at line level when you change product_id, to cover both cases: create and write.